Spectra

EthanDevlinSTFC opened the pull request:
To be rebased onto previous pr.
Implement Spectra plotting functionality
Fix context menu/ cycleMap issues
closes #99

trisyoungs reviewed (changes requested):
Yep, works great!  There's some clarification we should make to exactly what's being plotted - we have detectors and monitors, but while all monitors are also detectors they don't appear in the "detector_1" array. They get shoved into a separate thing for monitors.

1) In the menu let's rename "Plot spectra" to "Plot detector spectrum"
2) Make the title of the dialog "Plot Detector Spectrum" in line with the above
3) For the text in the dialog I think we should be explicit and say "Enter detector spectrum to plot (N.B. array indexing starts from first detector, excluding monitors):"
4) Can we determine the size of the array before the dialog is shown so we can clamp the input range to the available range? Could use `QInputDialog::getInt`.
5) How difficult is it to have an option on the plot widget to divide the displayed counts by the time-of-flight histogram binwidths? If that could be worked into the system, rather than having a separate menu option for it, that would be great.
6) The ability to "normalise" the counts by the collected beam current would also be a life-saver, but I guess arguably this could be a separate menu item.  Might be worth considering making the "Plot spectrum" a submenu with "Single detector", "Single monitor" options etc. Extensibility and all that!

Feel free to split this into smaller PRs if you deem it best.

EthanDevlinSTFC committed:
Bug fixes (#100)

* Fix changing cycle while grouped bug

* improve filter speed, halt unneccesary processes

* Fix run excess by showing title in status

* fix menu name issue

* broad cycle button implementation

* Cycle on button not combo

* go to fix

* mass search ux improvement

* remove debugging

* format

* context menu fix

* syntax fix

improve filter speed, halt unneccesary processes

Fix run excess by showing title in status

fix menu name issue

broad cycle button implementation

Cycle on button not combo

go to fix

mass search ux improvement

remove debugging

format

syntax fix

context menu fix

spectra

Plot single spectra full implementation

Spectra qol

refactor and rename

remove unused libraries

format

non-normalisation suggested changes

format

EthanDevlinSTFC committed:
Merge branch 'develop' into spectra

EthanDevlinSTFC committed:
format

trisyoungs commented on backend/isisInternal.py:
```diff
@@ -208,6 +208,22 @@ def getGoToCycle(instrument, search):
+# Get spectra data
+
+
+@app.route('/getSpectra/<instrument>/<cycle>/<runs>/<spectra>')
+def getSpectra(instrument, cycle, runs, spectra):
+    data = nexusInteraction.spectra(instrument, cycle, runs, spectra)
+    return jsonify(data)
```
If this function just returns a single spectrum then it should be called `getSpectrum`.

trisyoungs commented on backend/isisInternal.py:
```diff
@@ -208,6 +208,22 @@ def getGoToCycle(instrument, search):
+@app.route('/getSpectra/<instrument>/<cycle>/<runs>')
+def spectraRange(instrument, cycle, runs):
+    data = nexusInteraction.spectraRange(instrument, cycle, runs)
+    return jsonify(data)
```
For consistency and clarity I would be tempted to rename this to `getSpectrumRange`.

trisyoungs commented on backend/nexusInteraction.py:
```diff
@@ -108,3 +108,25 @@ def fieldData(instrument, cycle, runs, fields):
     return data
+
+
+def spectra(instrument, cycle, runs, spectra):
```
`getSpectrum`

trisyoungs commented on backend/nexusInteraction.py:
```diff
@@ -108,3 +108,25 @@ def fieldData(instrument, cycle, runs, fields):
+    return data
+
+
+def spectraRange(instrument, cycle, runs):
```
`getSpectrumRange`

trisyoungs commented on frontend/nexusInteraction.cpp:
```diff
@@ -420,4 +424,185 @@ void MainWindow::showStatus(qreal x, qreal y, QString title)
+        auto *gridLayout = new QGridLayout(window);
+
+        gridLayout->addWidget(chartView, 1, 0, -1, -1);
+        QString tabName = "Spectra";
```
We can make the tab name more informative - `Detector 1234` for example.

trisyoungs commented on frontend/nexusInteraction.cpp:
```diff
@@ -420,4 +424,185 @@ void MainWindow::showStatus(qreal x, qreal y, QString title)
+        }
+        chart->createDefaultAxes();
+        chart->axes(Qt::Horizontal)[0]->setTitleText("Time of flight");
+        chart->axes(Qt::Vertical)[0]->setTitleText("Bin value");
```
Instead of "Bin value" call this "Counts". When we add the option to divide through by the time-of-flight binwidth it will be "Counts/us" (i.e. counts per microsecond).

trisyoungs commented on frontend/nexusInteraction.cpp:
```diff
@@ -420,4 +424,185 @@ void MainWindow::showStatus(qreal x, qreal y, QString title)
+            chart->addSeries(series);
+        }
+        chart->createDefaultAxes();
+        chart->axes(Qt::Horizontal)[0]->setTitleText("Time of flight");
```
```suggestion
        chart->axes(Qt::Horizontal)[0]->setTitleText("Time of flight, &#181;s");
```
The `&#181;` should give the Greek mu.

trisyoungs commented on frontend/nexusInteraction.cpp:
```diff
@@ -420,4 +424,185 @@ void MainWindow::showStatus(qreal x, qreal y, QString title)
+    }
+}
+
+void MainWindow::getSpectraCount()
```
`getSpectrumCount`

trisyoungs commented on frontend/nexusInteraction.cpp:
```diff
@@ -420,4 +424,185 @@ void MainWindow::showStatus(qreal x, qreal y, QString title)
+    spectraCount.chop(1);
+    auto spectrumNumber =
+        QInputDialog::getInt(this, tr("Plot Detector Spectrum"),
+                             tr("Enter detector spectrum to plot (0-" + spectraCount + "):"), 0, 0, count->response.toInt(), 1);
```
Arrays in the NeXuS file are zero-indexed I believe, so the max value should be `count->response.toInt() - 1`

trisyoungs commented on frontend/nexusInteraction.cpp:
```diff
@@ -420,4 +424,185 @@ void MainWindow::showStatus(qreal x, qreal y, QString title)
+    // Gets all selected run numbers and fills graphing toggles
+    QString runNos = "";
+    QString runNo;
+    // Concats runs
+    for (auto run : selectedRuns)
+    {
+        runNo = model_->index(run.row(), runNoColumn).data().toString();
+        if (runNo.contains("-") || runNo.contains(","))
+        {
+            QString groupedRuns;
+            auto runArr = runNo.split(",");
+            foreach (const auto &string, runArr)
+            {
+                if (string.contains("-"))
+                {
+                    for (auto i = string.split("-")[0].toInt(); i <= string.split("-")[1].toInt(); i++)
+                        groupedRuns += QString::number(i) + ";";
+                }
+                else
+                    groupedRuns += string + ";";
+            }
+            groupedRuns.chop(1);
+            runNos.append(groupedRuns + ";");
+        }
+        else
+            runNos.append(runNo + ";");
+    }
```
This chunk of code looks identical to that in the routine above - would be worth splitting this out into a private function to save on code duplication.

trisyoungs reviewed (approved):
:+1: Handful of comments, mostly in regard to naming of stuff, and one small bugfix.

EthanDevlinSTFC committed:
suggested changes
